File: HW10/bipartite.py
def fulkerson(graph, s, t):
    def dfs(graph, s, t, visited, flow):
        visited[s] = True
        if s == t:
            return flow
        for i in range(len(graph)):
            if not visited[i] and graph[s][i] > 0:
                new_flow = min(flow, graph[s][i])
                returned_flow = dfs(graph, i, t, visited, new_flow)
                if returned_flow > 0:
                    graph[s][i] -= returned_flow
                    graph[i][s] += returned_flow
                    return returned_flow
        return 0

    max_flow = 0
    while True:
        visited = [False] * len(graph)
        flow = dfs(graph, s, t, visited, float("inf"))
        if flow == 0:
            break
        max_flow += flow

    return max_flow

t = int(input())
for _ in range(t):
    m, n, edges = [int(num) for num in input().split(" ")]

    graph = [[0] * (m+n+2) for _ in range(m+n+2)]
    flow = [[0] * (m+n+2) for _ in range(m+n+2)]
    
    graph[0] = [0] + [1] * m + [0] * (n+1)
    for i in range(n):
        graph[m+1+i][-1] = 1
    
    for i in range(edges):
        e = [int(num) for num in input().split(" ")]

        graph[e[0]][e[1]+m] = 1

    max_flow = fulkerson(graph, 0, len(graph)-1)
    print(max_flow, "Y" if max_flow == n and max_flow == m else "N")

File: HW10/bipartite.rs
use std::io;
fn dfs(graph: &mut Vec<Vec<i32>>, s: usize, t: usize, visited: &mut Vec<bool>, flow: i32) -> i32 {
    visited[s] = true;
    if s == t {
        return flow;
    }
    for i in 0..graph.len() {
        if !visited[i] && graph[s][i] > 0 {
            let new_flow = flow.min(graph[s][i]);
            let returned_flow = dfs(graph, i, t, visited, new_flow);
            if returned_flow > 0 {
                graph[s][i] -= returned_flow;
                graph[i][s] += returned_flow;
                return returned_flow;
            }
        }
    }
    return 0;
}

fn max_flow(graph: &mut Vec<Vec<i32>>, s: usize, t: usize) -> i32 {
    let mut max_flow = 0;
    loop {
        let mut visited = vec![false; graph.len()];
        let flow = dfs(graph, s, t, &mut visited, std::i32::MAX);
        if flow == 0 {
            break;
        }
        max_flow += flow;
    }
    return max_flow;
}



fn main() {
    let mut buff = String::new();
    io::stdin().read_line(&mut buff).unwrap();

    let n_instances = buff.trim().parse().unwrap();
    buff.clear();

    for _ in 0..n_instances {
        io::stdin().read_line(&mut buff).unwrap();
        let size: Vec<usize> = buff.trim().split(" ")
                .map(|x| x.parse().unwrap())
                .collect();
        buff.clear();

        let vector_len = size[0] + size[1] + 2;
        let mut graph = vec![vec![0; vector_len]; vector_len];

        for i in 1..size[0]+1 {
            graph[0][i] = 1;
        }

        for i in 0..size[1] {
            graph[size[0]+1+i][vector_len-1] = 1
        }

        for _ in 0..size[2] {
            io::stdin().read_line(&mut buff).unwrap();
            let e: Vec<usize> = buff.trim().split(" ")
                .map(|x| x.parse().unwrap())
                .collect();
            buff.clear();

            graph[e[0]][e[1]+size[0]] = 1
        }

        let f: i32 = max_flow(&mut graph, 0, vector_len-1);
        println!("{} {}", f, if f == size[1] as i32 && f == size[0] as i32 {"Y"} else {"N"});
    }
}